accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/helper/AceBean.java

```diff
@@ -249,47 +259,101 @@ private void removeRedundantPrivileges(Session session) throws RepositoryExcepti
+		Collection<String> supportedRestrictionNames = Arrays.asList(acl.getRestrictionNames());
+		if (getRepGlob() != null) {
+			if (!supportedRestrictionNames.contains(RESTRICTION_NAME_GLOB)) {
+				throw new IllegalStateException("The AccessControlList at " + acl.getPath() + " does not support setting rep:glob restrictions!");
+			}
+			Value v = session.getValueFactory().createValue(getRepGlob(), acl.getRestrictionType(RESTRICTION_NAME_GLOB));
+			return Collections.singletonMap(RESTRICTION_NAME_GLOB, v);
+		} else {
+			return Collections.emptyMap();
+		}
+	}
+ 	
+	/**
+	 * Creates an action map being used in {@link CqActions#installActions(String, Principal, Map, Collection)} out of the set actions on this bean.
+	 * @return a map containing actions as keys and booleans representing {@code true} for allow and {@code false} for deny.
+	 */
+	private Map<String, Boolean> getActionMap() {
+		if (actions == null) {
+			return Collections.emptyMap();
+		}
+		Map<String, Boolean> actionMap = new HashMap<String, Boolean>();
+		for (String action : actions) {
+            actionMap.put(action, isAllow());
         }
-        
-        
-		JackrabbitAccessControlList acl = AccessControlUtils.getModifiableAcl(
-				acMgr, getJcrPath());
-		Set<Privilege> privileges = AccessControlUtils.getPrivilegeSet(
-				getPrivileges(), acMgr);
+		return actionMap;
+	}
+
+	/**
+	 * Installs the CQ actions in the repository.
+	 * 
+	 * @param principal
+	 * @param acl
+	 * @param session
+	 * @param acMgr
+	 * @return either the same acl as given in the parameter {@code acl} if no actions have been installed otherwise the new AccessControlList (comprising the entres being installed for the actions).
+	 * @throws RepositoryException
+	 */
+	private JackrabbitAccessControlList installActions(Principal principal, JackrabbitAccessControlList acl, Session session, AccessControlManager acMgr) throws RepositoryException {
+		Map<String, Boolean> actionMap = getActionMap();
+		if (actionMap.isEmpty()) {
+			return acl;
+		}
+		int previousAclSize = acl.size();
 		
-		if (!privileges.isEmpty()) {
-			Map<String, Value> restrictions = null;
-			if (getRepGlob() != null) {
-				// is rep:glob supported?
-				for (String rName : acl.getRestrictionNames()) {
-					if ("rep:glob".equals(rName)) {
-						Value v = session.getValueFactory().createValue(
-								getRepGlob(), acl.getRestrictionType(rName));
-						restrictions = Collections.singletonMap(rName, v);
-						break;
-					}
-				}
+		CqActions cqActions = new CqActions(session);
+		Collection<String> inheritedAllows = cqActions.getAllowedActions(
+				getJcrPath(), Collections.singleton(principal));
+		// this does always install new entries
+		cqActions.installActions(getJcrPath(), principal, actionMap,
+				inheritedAllows);
+		
```

Comment:
Since there are potential changes in the acl done by the call to cqActions.installActions(), the acl has to be retrieved again directly afterwards. Otherwise the old ace gets returned in case only actions (without globbing, or additional privileges) would have been set. So, the later call to: AccessControlUtils.getAccessControlList(session, getJcrPath()) has to be moved up.